includes/class-external-token.php

```diff
@@ -0,0 +1,314 @@
+	 */
+	public function update( $info ) {
+		if ( ! array_key_exists( 'access_token', $info ) ) {
+			return false;
+		}
+
+		if ( array_key_exists( 'expires_in', $info ) && is_numeric( $info['expires_in'] ) && ! array_key_exists( 'expiration', $info ) ) {
+			$info['expiration'] = time() + $info['expires_in'];
+		}
+
+		$tokens = get_user_meta( $this->user_id, 'indieauth_external_tokens', true );
+		if ( ! is_array( $tokens ) ) {
+			$tokens = array();
+		}
+
+		$found = null;
+		foreach ( $tokens as $key => $token ) {
+			if ( in_array( $info['access_token'], $token, true ) ) {
+				$found = $key;
+			}
+		}
+		if ( ! $found ) {
+			// Add local unique identifier to distinguish if multiple tokens are issued.
+			$info['uuid'] = wp_generate_uuid4();
+
+			$tokens[] = $info;
+		} else {
+			$tokens[ $found ] = array_merge( $tokens[ $found ], $info );
+		}
+
+		return update_user_meta( $this->user_id, 'indieauth_external_tokens', $tokens );
+	}
+
+	/**
+	 * Destroys a token.
+	 *
+	 * @param array|string $tokens Token to destroy. Will also accept an array of tokens to destroy.
+	 * @param boolean $revoke Whether to send revoke request to token endpoint.
+	 */
+	public function destroy( $destroy, $revoke = true ) {
+		if ( ! current_user_can( 'edit_user', $this->user_id ) ) {
+			return false;
+		}
+
+		$tokens = get_user_meta( $this->user_id, 'indieauth_external_tokens', true );
+
+		if ( is_string( $destroy ) ) {
+			$destroy = array( $destroy );
+		}
+
+		foreach ( $tokens as $key => $token ) {
+			if ( in_array( $token['access_token'], $destroy, true ) ) {
+				unset( $tokens[ $key ] );
+				if ( $revoke ) {
+					$resp = $this->revoke_external_token( $token );
+				}
+			}
+		}
+
+		update_user_meta( $this->user_id, 'indieauth_external_tokens', $tokens );
+	}
+
+	/**
+	 * Revokes an External token.
+	 *
+	 * @param array $token Token to destroy. This is the token info stored in the database.
+	 * @param boolean $revoke Whether to send revoke request to token endpoint.
+	 * @return boolean|array Either false or the response from the token endpoint.
+	 */
+	protected function revoke_external_token( $token ) {
+		// Use stored token endpoint or discover it.
+		if ( array_key_exists( 'token_endpoint', $token ) ) {
+			$token_endpoint = $token['token_endpoint'];
+		} else {
+			$token_endpoint = find_rels( $token['resource'], 'token_endpoint' );
+		}
+
+		if ( ! wp_http_validate_url( $token_endpoint ) ) {
+			return false;
+		}
+
+		$resp = wp_safe_remote_post(
+			$token_endpoint,
+			array(
+				'headers' => array(
+					'Accept' => 'application/json',
+				),
+				'body'    => array(
+					'action' => 'revoke',
+					'token'  => $token['access_token'],
+				),
+			)
+		);
+		return $resp;
+	}
+
+
+	/**
+	 * Revokes an External token.
+	 *
+	 * @param array $key Token to verify.
+	 * @param boolean $revoke Whether to send revoke request to token endpoint.
+	 * @return boolean|array Either false or the response from the token endpoint.
```

Comment:
Minor nitpick, but this feels like it doesn't revoke, just verify. Plus there's the `$revoke` parameter that doesn't exist on this `verify()` method

Reply:
Fixing that now. I think I just used the docblock from another function and forgot to fix it.